ntbs-integration-tests/Helpers/WebApplicationExtensions.cs

```diff
@@ -13,41 +12,46 @@ namespace ntbs_integration_tests.Helpers
         {
             return factory.WithWebHostBuilder(builder =>
             {
-                SetupDatabase(builder, idServiceCodeMapping);
                 builder.ConfigureTestServices(services =>
                 {
-                    services.AddScoped<IUserService, NhsUserService>();
+                    services.AddScoped<IUserService, T>();
                 });
             });
         }
 
-        public static WebApplicationFactory<Startup> WithPheUserBuilder(this WebApplicationFactory<Startup> factory,
-                                                                        IDictionary<int, string> idServiceCodeMapping,
-                                                                        IDictionary<int, string> idPostcodeMapping)
+        public static WebApplicationFactory<Startup> WithNotificationAndTbServiceConnected(this WebApplicationFactory<Startup> factory,
+                                                                                            int notificationId,
+                                                                                            string tbServiceCode)
         {
             return factory.WithWebHostBuilder(builder =>
             {
-                SetupDatabase(builder, idServiceCodeMapping, idPostcodeMapping);
-                builder.ConfigureTestServices(services =>
-                {
-                    services.AddScoped<IUserService, PheUserService>();
-                });
+                SetupDatabase(builder, new Tuple<int, string>(notificationId, tbServiceCode), null);
+            });
+        }
+
+        public static WebApplicationFactory<Startup> WithNotificationAndPostcodeConnected(this WebApplicationFactory<Startup> factory,
+                                                                                            int notificationId,
+                                                                                            string postcode)
+        {
+            return factory.WithWebHostBuilder(builder =>
+            {
+                SetupDatabase(builder, null, new Tuple<int, string>(notificationId, postcode));
             });
         }
 
-        public static HttpClient WithoutRedirects(this WebApplicationFactory<Startup> factory)
+        public static HttpClient CreateClientWithoutRedirects(this WebApplicationFactory<Startup> factory)
         {
             return factory.CreateClient(new WebApplicationFactoryClientOptions
             {
                 AllowAutoRedirect = false
             });
         }
 
-        private static void SetupDatabase(IWebHostBuilder builder, IDictionary<int, string> idServiceCodePairs, IDictionary<int, string> idPostcodePairs = null)
+        private static void SetupDatabase(IWebHostBuilder builder, Tuple<int, string> idToServiceCode, Tuple<int, string> idToPostcode)

```

Comment:
This is much better I think, but we still haven't eradicated these awkward (because of how specific they are) tuples completely. Would it make more sense to pass a list of lambdas into this method, which the method would execute on the `db` object?

Reply:
Yes that's a good idea! I was thinking as this method isn't really gonna be viewed by people when people are writing tests it's probably fine with the tuples, but obviously if we want to add more db update methods in future it isn't as extensible